File: examples/go/player/agg/handlers/release.go
package handlers

import (
	"encoding/hex"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardReleaseFunds(state PlayerState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Player does not exist")
	}
	return nil
}

func validateReleaseFunds(cmd *examples.ReleaseFunds, state PlayerState) (int64, error) {
	if cmd.TableRoot == nil || len(cmd.TableRoot) == 0 {
		return 0, angzarr.NewCommandRejectedError("table_root is required")
	}

	tableKey := hex.EncodeToString(cmd.TableRoot)
	reserved, ok := state.TableReservations[tableKey]
	if !ok {
		return 0, angzarr.NewCommandRejectedError("No funds reserved for this table")
	}

	return reserved, nil
}

func computeFundsReleased(cmd *examples.ReleaseFunds, state PlayerState, reserved int64) *examples.FundsReleased {
	newReserved := state.ReservedFunds - reserved
	newAvailable := state.Bankroll - newReserved
	return &examples.FundsReleased{
		Amount:              &examples.Currency{Amount: reserved, CurrencyCode: "CHIPS"},
		TableRoot:           cmd.TableRoot,
		NewAvailableBalance: &examples.Currency{Amount: newAvailable, CurrencyCode: "CHIPS"},
		NewReservedBalance:  &examples.Currency{Amount: newReserved, CurrencyCode: "CHIPS"},
		ReleasedAt:          timestamppb.New(time.Now()),
	}
}

// HandleReleaseFunds handles the ReleaseFunds command (release reserved funds back to bankroll).
func HandleReleaseFunds(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state PlayerState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.ReleaseFunds
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardReleaseFunds(state); err != nil {
		return nil, err
	}
	reserved, err := validateReleaseFunds(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computeFundsReleased(&cmd, state, reserved)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}

File: examples/rust/player/agg/src/handlers/release.rs
//! ReleaseFunds command handler.

use angzarr_client::proto::examples::{Currency, FundsReleased, ReleaseFunds};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::PlayerState;

pub fn handle_release_funds(
    command_book: &CommandBook,
    command_any: &Any,
    state: &PlayerState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Player does not exist"));
    }

    let cmd: ReleaseFunds = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    if cmd.table_root.is_empty() {
        return Err(CommandRejectedError::new("table_root is required"));
    }

    let table_key = hex::encode(&cmd.table_root);
    let reserved = match state.table_reservations.get(&table_key) {
        Some(&amount) => amount,
        None => return Err(CommandRejectedError::new("No funds reserved for this table")),
    };

    let new_reserved = state.reserved_funds - reserved;
    let new_available = state.bankroll - new_reserved;

    let event = FundsReleased {
        amount: Some(Currency {
            amount: reserved,
            currency_code: "CHIPS".to_string(),
        }),
        table_root: cmd.table_root,
        new_available_balance: Some(Currency {
            amount: new_available,
            currency_code: "CHIPS".to_string(),
        }),
        new_reserved_balance: Some(Currency {
            amount: new_reserved,
            currency_code: "CHIPS".to_string(),
        }),
        released_at: Some(angzarr_client::now()),
    };

    let event_any = pack_event(&event, "examples.FundsReleased");

    Ok(new_event_book(command_book, seq, event_any))
}
